cli/src/diff_util.rs

```diff
@@ -400,83 +401,87 @@ fn basic_diff_file_type(values: &MergedTreeValue) -> String {
+    mut tree_diff: TreeDiffStream,
 ) -> Result<(), CommandError> {
     let repo = workspace_command.repo();
     formatter.push_label("diff")?;
-    for (path, diff) in tree_diff {
-        let ui_path = workspace_command.format_file_path(&path);
-        let (left_value, right_value) = diff?;
-        if left_value.is_absent() {
-            let right_content = diff_content(repo, &path, &right_value)?;
-            let description = basic_diff_file_type(&right_value);
-            writeln!(
-                formatter.labeled("header"),
-                "Added {description} {ui_path}:"
-            )?;
-            if right_content.is_empty() {
-                writeln!(formatter.labeled("empty"), "    (empty)")?;
-            } else {
-                show_color_words_diff_hunks(&[], &right_content, formatter)?;
-            }
-        } else if right_value.is_present() {
-            let left_content = diff_content(repo, &path, &left_value)?;
-            let right_content = diff_content(repo, &path, &right_value)?;
-            let description = match (left_value.into_resolved(), right_value.into_resolved()) {
-                (
-                    Ok(Some(TreeValue::File {
-                        executable: left_executable,
-                        ..
-                    })),
-                    Ok(Some(TreeValue::File {
-                        executable: right_executable,
-                        ..
-                    })),
-                ) => {
-                    if left_executable && right_executable {
-                        "Modified executable file".to_string()
-                    } else if left_executable {
-                        "Executable file became non-executable at".to_string()
-                    } else if right_executable {
-                        "Non-executable file became executable at".to_string()
-                    } else {
-                        "Modified regular file".to_string()
-                    }
-                }
-                (Err(_), Err(_)) => "Modified conflict in".to_string(),
-                (Err(_), _) => "Resolved conflict in".to_string(),
-                (_, Err(_)) => "Created conflict in".to_string(),
-                (Ok(Some(TreeValue::Symlink(_))), Ok(Some(TreeValue::Symlink(_)))) => {
-                    "Symlink target changed at".to_string()
-                }
-                (Ok(left_value), Ok(right_value)) => {
-                    let left_type = basic_diff_file_type(&Merge::resolved(left_value));
-                    let right_type = basic_diff_file_type(&Merge::resolved(right_value));
-                    let (first, rest) = left_type.split_at(1);
-                    format!(
-                        "{}{} became {} at",
-                        first.to_ascii_uppercase(),
-                        rest,
-                        right_type
-                    )
+    async {
+        while let Some((path, diff)) = tree_diff.next().await {
```

Comment:
> I'm going to implement a `Stream`-based version optimized for
high-latency (RPC-based) commit backends. So far, that implementation
is about 20% slower in the Linux repo when running `jj diff
--ignore-working-copy -s --from v5.0 --to v6.0`. I think that's almost
only because the algorithm is different, not because it's async per
se.
>
>This commit adds a `Stream`-based version of `MergedTree::diff()` that
just wraps the regular iterator in stream. I updated `jj diff` to use
it. I couldn't measure any difference on the command above in the
Linux repo. I think that means we can safely use the same
`Stream`-based interface regardless of backend, even if we end up
needing two different implementations of the `Stream`. We would then
be using the wrapped iterator from this commit for local backends, and
the new implementation for remote backends. But ideally we can make
the remote-friendly implementation fast enough that we don't need two
implementations.

Driveby: I'd put some parts of the motivation in a comment above the block, as not everyone looks at commit messages.

Reply:
I'll merge this PR. If you prefer that I move the comment somewhere else, just let me know and I'll do it in a followup.